core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java

```diff
@@ -0,0 +1,296 @@
+    opts.parseArgs(GenerateSplits.class.getName(), args);
+    if (opts.files.isEmpty()) {
+      throw new IllegalArgumentException("No files were given");
+    }
+
+    Configuration hadoopConf = new Configuration();
+    SiteConfiguration siteConf = opts.getSiteConfiguration();
+    boolean encode = opts.base64encode;
+
+    TreeSet<String> splits;
+    TreeSet<String> desiredSplits = new TreeSet<>();
+
+    if (opts.numSplits > 0 && opts.splitSize > 0) {
+      throw new IllegalArgumentException("Requested number of splits and split size.");
+    }
+    if (opts.numSplits == 0 && opts.splitSize == 0) {
+      throw new IllegalArgumentException("Required number of splits or split size.");
+    }
+    long numSplits = opts.numSplits;
+    long splitSize = opts.splitSize;
+
+    FileSystem fs = FileSystem.get(hadoopConf);
+    FileSystem localFs = FileSystem.getLocal(hadoopConf);
+    List<Path> filePaths = new ArrayList<>();
+    for (String file : opts.files) {
+      Path path = new Path(file);
+      if (file.contains(":")) {
+        fs = path.getFileSystem(hadoopConf);
+      } else {
+        log.warn("Attempting to find file across filesystems. Consider providing URI "
+            + "instead of path");
+        if (!fs.exists(path))
+          fs = localFs; // fall back to local
+      }
+      // get all the files in the directory
+      if (fs.getFileStatus(path).isDirectory()) {
+        // can only explode one directory
+        if (opts.files.size() > 1)
+          throw new IllegalArgumentException("Only one directory can be specified");
+        var iter = fs.listFiles(path, true);
+        while (iter.hasNext()) {
+          filePaths.add(iter.next().getPath());
+        }
+      } else {
+        filePaths.add(path);
+      }
+    }
+
+    // if no size specified look at indexed keys first
+    if (opts.splitSize == 0) {
+      splits = getIndexKeys(siteConf, hadoopConf, fs, filePaths, encode);
+      // if there weren't enough splits indexed, try again with size = 0
+      if (splits.size() < numSplits) {
+        log.info("Only found {} indexed keys but need {}. Doing a full scan on files {}",
+            splits.size(), numSplits, filePaths);
+        splits = getSplitsBySize(siteConf, hadoopConf, filePaths, fs, 0, encode);
+      }
+    } else {
+      splits = getSplitsBySize(siteConf, hadoopConf, filePaths, fs, splitSize, encode);
+    }
+
+    int numFound = splits.size();
+    // its possible we found too many indexed so take every (numFound / numSplits) split
+    if (opts.splitSize == 0 && numFound > numSplits) {
+      var iter = splits.iterator();
+      long factor = numFound / numSplits;
+      log.debug("Found {} splits but requested {} picking 1 every {}", numFound, opts.numSplits,
+          factor);
+      for (int i = 0; i < numFound; i++) {
+        String next = iter.next();
+        if (i % factor == 0 && desiredSplits.size() < numSplits) {
+          desiredSplits.add(next);
+        }
+      }
```

Comment:
For a case where numFound is 150 and numSplits is 100, 150/100 is 1, so factor is 1.   `i % 1` is always zero, so I think the first 100 splits out of the 150 will be taken.  This probably not desired.   Thinking an approach similar to the following will yield better results.  Hopefully the following code would take splits `0,1,3,4,6,7,9,10, etc` but I am not sure its correct... however that is the intent.

```suggestion
      double targetFactor = (double)numFound / numSplits;
      log.debug("Found {} splits but requested {} picking 1 every {}", numFound, opts.numSplits,
          factor);
  
      for (int i = 0; i < numFound; i++) {
        double currFactor = (double)(i+1)/desiredSplits.size();
        String next = iter.next();
        // unsure if this should be currFactor >= targetFactor .. need to think through some edge cases
        if (currFactor > targetFactor && desiredSplits.size() < numSplits) {
          desiredSplits.add(next);
        }
      }
```

Reply:
This code also takes a subset of splits using floating point arithmetic. https://github.com/apache/accumulo/blob/45a4a93d2926760d6a916ae2da1d7e668ec4dcb1/core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java#L693-L714

Reply:
I think the code I just linked has bug, will open an issue.

Reply:
I opened #2371 for a possible bug w/ the listSplits table operation. Thinking if there is a slight bug w/ that code that it could possibly use the approach above, but not sure.

Reply:
I tested your scenario and you are correct:
<pre>
2021-11-29T14:36:00,483 [rfile.GenerateSplits] DEBUG: Found 150 splits but requested 100 picking 1 every 1
2021-11-29T14:36:00,484 [rfile.GenerateSplits] INFO : Generated 100 splits
row_0000000000
row_0000000001
row_0000000002
row_0000000003
row_0000000004
...
</pre>
It will then select the first 100 splits it sees. I will do some testing of your suggestion for using two floating point values for measuring the splits.